quic-client/src/lib.rs

```diff
@@ -158,65 +165,123 @@ impl QuicConfig {
     }
 
     fn new_nonblocking_connection(
         &self,
         _addr: SocketAddr,
         stats: Arc<ConnectionCacheStats>,
-    ) -> NonblockingQuicTpuConnection {
-        NonblockingQuicTpuConnection::new_with_client(self.0.clone(), stats)
+    ) -> Arc<dyn NonblockingClientConnection> {
+        Arc::new(NonblockingQuicClientConnection::new_with_client(
+            self.0.clone(),
+            stats,
+        ))
+    }
+}
+
+#[derive(Default)]
+pub struct QuicConnectionManager {
+    connection_config: Option<Box<dyn NewConnectionConfig>>,
+}
+
+impl ConnectionManager for QuicConnectionManager {
+    fn new_connection_pool(&self) -> Box<dyn ConnectionPool> {
+        Box::new(QuicPool {
+            connections: Vec::default(),
+            endpoint: Arc::new(self.connection_config.as_ref().map_or(
+                QuicLazyInitializedEndpoint::default(),
+                |config| {
+                    let config: &QuicConfig = match config.as_any().downcast_ref::<QuicConfig>() {
+                        Some(b) => b,
+                        None => panic!("Expecting a QuicConfig!"),
+                    };
+
+                    config.create_endpoint()
+                },
+            )),
+        })
+    }
+
+    fn new_connection_config(&self) -> Box<dyn NewConnectionConfig> {
+        Box::new(QuicConfig::new().unwrap())
+    }
+
+    fn get_port_offset(&self) -> u16 {
+        QUIC_PORT_OFFSET
+    }
+
+    fn get_protocol_type(&self) -> ProtocolType {
+        ProtocolType::QUIC
     }
 }
 
+impl QuicConnectionManager {
+    pub fn new_with_connection_config(config: QuicConfig) -> Self {
+        Self {
+            connection_config: Some(Box::new(config)),
+        }
+    }
+}
 #[cfg(test)]
 mod tests {
     use {
         super::*,
+        solana_connection_cache::connection_cache::{
+            ConnectionCache, DEFAULT_CONNECTION_POOL_SIZE,
+        },
         solana_sdk::quic::{
             QUIC_MAX_UNSTAKED_CONCURRENT_STREAMS, QUIC_MIN_STAKED_CONCURRENT_STREAMS,
             QUIC_TOTAL_STAKED_CONCURRENT_STREAMS,
         },
-        solana_tpu_client::tpu_connection_cache::{
-            TpuConnectionCache, DEFAULT_TPU_CONNECTION_POOL_SIZE,
-        },
     };
 
     #[test]
     fn test_connection_cache_max_parallel_chunks() {
         solana_logger::setup();
+        let connection_manager = Box::<QuicConnectionManager>::default();
         let connection_cache =
-            TpuConnectionCache::<QuicPool>::new(DEFAULT_TPU_CONNECTION_POOL_SIZE).unwrap();
-        let mut tpu_config = connection_cache.tpu_config;
+            ConnectionCache::new(connection_manager, DEFAULT_CONNECTION_POOL_SIZE).unwrap();
+        let mut connection_config = connection_cache.connection_config;
+
+        let connection_config: &mut QuicConfig =
+            match connection_config.as_mut_any().downcast_mut::<QuicConfig>() {
+                Some(b) => b,
+                None => panic!("Expecting a QuicConfig!"),
+            };
```

Comment:
This feels fairly awful to use. Can we add a helper fn on QuicConfig or freestanding in this crate to do this? (and another for the non-mut version)

Reply:
Will do

Reply:
Done